### PR TITLE
Final tutorial submission: Automatic configuration of Kubernetes using Ansible

### PR DESCRIPTION
**Pull request by fwallb**
# Executable tutorial: *Automatic configuration of Kubernetes using Ansible*

### Members
Matej Sestak (sestak@kth.se)
GitHub: [sestys](https://github.com/sestys)

Frida Wallberg (fwallb@kth.se)
GitHub: [fwallb](https://github.com/fwallb)

## Description
A key aspect of DevOps is automation. Ansible is used for automation of deployment and configuration management. Kubernetes is used for container orchestration. Together they can be used to automate container orchestration.
The tutorial gives an introduction to Ansible and Kubernetes (K8s), and how work together. I touches upon Ansible playbooks and inventory files, as well as Kubernetes clusters and namespaces. Lastly Ansible is used to automate the process of deploying a Docker image to Kubernetes.

## Structure of the tutorial
* Introduction, overview of Ansible and K8s, how Ansible and K8s fit into the DevOps pipeline,
* Create a cluster,
* Ansible set up,
* Create an inventory file,
* Intoduction to Ansible playbooks,
* Create a simple playbook,
* K8s and namespaces,
* Pull a Docker image,
* Deploy the Docker image to K8s using Ansible,
* Did you find the *easter egg* 🐥?,
* Sum up and take away.

## Link to Katacoda scenario
[**Configuring K8s with Ansible**](https://www.katacoda.com/sestamat/scenarios/k8s-with-ansible)

## Link to Github
[**Source code for Katakoda scenario**](https://github.com/sestys/katacoda-scenarios)

## Aiming for distinction

|                                             | Yes | No | Remarkable |
|-------------------------------------------- | ----|----|-------------|
|The TA can successful execute all the commands of the tutorial (mandatory) |✅Yes|No|✅In the browser  |
|~~If local execution, runs on Linux~~ | ~~Yes~~ | ~~No~~ | ~~Easy to setup and run~~  |
|The tutorial gives enough background |✅Yes | No | ✅Comprehensive background |
|The tutorial is easy to follow  |✅Yes | No | ✅Well documented |
|The tutorial is original, no such tutorial exists on the web | ✅Yes | No | The teaching team never heard about it |
|The tutorial contains [easter eggs](https://github.com/OrkoHunter/python-easter-eggs) |✅Yes|No| ✅Subtle and fun |
|The tutorial is successful (attracts comments and success) |✅Yes | No | Lively discussion |
|The language is correct |✅Yes | No |Interesting narrative  |

### CODE REVIEW

**Comment by fwallb** [comment]
@Wizkas0, the tutorial is now submitted and we are looking forward to your feedback!

**Comment by Wizkas0** [comment]
Great! I'll get on it!

**Comment by Wizkas0** [comment]
Here is my feedback, hope it helps you! I truly enjoyed your tutorial!

# Feedback
## Overall thoughts
Overall, I thought the tutorial was of great quality. It was easy to follow, and it taught me exactly what was advertised. The tutorial made the topic feel interesting, something that cannot be said for some  of tutorials out there. I feel this was because the purpose of each step was clearly communicated, so it never felt like I was just following instructions without knowing what affect they were having. The usefulness of each tool and method, in the wider scope of continuous deployment, was always made clear. All the parts were woven together seamlessly into the whole, in a way that felt natural and comprehensible.

I really liked how the steps involving the playbook built upon each other in a sensible way, at each step adding a new feature, then running it again. Doing it multiple times, each time slightly altering the playbook made me comfortable with the procedure and made me appreciate what each part of the playbook did.

My only overall gripe with the tutorial is the estimated length. It took me about double the suggested time to complete the tutorial (30min), and while that might be because I was trying to be extra thorough, I feel that more than fifteen minutes are required to truly comprehend the material. I do, however, have some suggestions for small improvements for some of the steps, which I will lay out below.

## Suggestions

### Introduction
The sentence `“In one of the two parts that form DevOps, operations, continuous deployment (CD) is the core principle.”` is unwieldy, so you may want to change it to `“Operations is one of the two parts that form DevOps, and its core principle is continuous deployment.“`. Also, I would suggest bumping up the estimated time to 25 min, since that is closer to my experience, and since the tutorial is aimed at beginners.

### Step 1
The first time I ran the `minikube start --wait=false` command, the process terminated after a few minutes with the output `killed`, which looks like another error than the one mentioned in the text. Rerunning fixed the error, just like the one mentioned in the text, so you should perhaps add to the instructions that the `killed` error can be solved in the same way.

### Step 3
Since the tutorial is aimed at beginners, it might be beneficial for them if you specify exactly what part of the output from the `kubectl cluster-info` command constitutes the clusters IP address. Though it is obvious to me what part is the IP and what part is the port number, to a complete beginner, it might not be. I suggest that you include an example output (just like you did for the hosts file) with the IP underlined, just to be sure (something like this: `Kubernetes master is running at https://_._._._:8443`). Furthermore, after running the `echo "[master]" > hosts` command, the hosts file would not show up until I refreshed the file tree. You should add instructions to press the `refresh tree` button at the top after running the command.

### Step 4
I believe the description for the `become` command could be extended, since this is the only command that I did not quite understand the purpose of. It would be great if you could explain what privilege escalation entails and what its purpose is.

### Step 5
I completed this tutorial two times, and the first time, after running the `ansible-playbook -i hosts playbook.yml` command, I did not receive the `ok=2` and `changed=1` message as expected. It seems the installation of yamllint failed, so I got to `ok=2` and `failed=1`. No step after this one failed however, and this step worked perfectly on my second run of the tutorial, so I am not quite sure what caused the error. Since this failure did not seem to affect the subsequent steps, perhaps this task could be removed or replaced with something less prone to failure, though this might just be user error, or something katacoda-related. Furthermore, just like in step 3, I had to refresh the file tree to be able to see the playbook file.

`I did not find any faults with the steps not mentioned.`

## Conclusion and easter egg
In conclusion, this was truly a great tutorial, with only some minor issues. I can add that I did find the easter egg, it was hidden in an ingenious way, both hard to find and very rewarding! You obviously worked hard on this tutorial and it shows. Bravo!

**Comment by fwallb** [comment]
@Wizkas0, thank you for the great feedback! ✨

**Comment by sestys** [comment]
@Wizkas0 Thank you for the feedback! We made the proposed changes, they were really concrete and actionable. We tried a few trial runs to check the problem in step 5 with yamllint and did not encounter the problem you described. Maybe there was some problem with Katacoda at the time. We decided to keep the yamllint in, as it is a useful tool to have installed, as it gives the reader possibility to check their playbook.yml file if they have correct formatting. 

Thanks one more for the awesome feedback!

**Comment by Wizkas0** [comment]
No problem! Yeah, I'm sure it was just some special case, may have had to do with my browser or something, just wanted to give you a heads up!

**Comment by SophieHYe** [comment]
Thanks for the final tutorial submission. I am now merging your PR and we will grade your work based on the final version.
